Removing waffle flag for responsive header.

HarryRein opened the pull request:
LEARNER-2955

This will make the responsive header ubiquitous across the LMS pages and assure that it is a11y compatible.

@andy-armstrong 
@AlasdairSwan 
@cptvitamin

AlasdairSwan commented:
The new header still looks quite different than the edx.org header, it would be great if you could update it to be more consistent.
<img width="720" alt="screen shot 2017-10-30 at 1 44 52 pm" src="https://user-images.githubusercontent.com/2052185/32186543-bb855018-bd78-11e7-86a2-dd0749645234.png">

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L12-L17
```.mobile-nav-item``` and ```.mobile-menu``` should be scoped to the header.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L25-L37

1. All selectors should be scoped to within the header.
2. ```.show``` and ```.hide``` should be replaced with adding/removing classes due to a performance issue with those jQuery functions.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L40-L51

1. All selectors should be scoped to within the header.
2. ```.show``` and ```.hide``` should be replaced with adding/removing classes due to a performance issue with those jQuery functions.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L55
jQuery should not be used for styling unless absolutely necessary, please update to add a class that will set to display none

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L61-L67

1. I would move this up to be in the on document ready and update it to only instantiate the listener if the header user menu is on the page.
2. ```.show``` and ```.hide``` should be replaced with adding/removing classes due to a performance issue with those jQuery functions.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L69-L70

This needs to be updated to reduce scope to just the mobile menu and user menu. I think there should be 2 separate event listeners, one for each. That will reduce the complexity of your nested if statements.

AlasdairSwan commented:
```KeyboardEvent.keyCode``` has been deprecated (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) please use ```KeyboardEvent.key``` instead (https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key)

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L87

It would be cleaner to programmatically close the menu (you could restructure your code to have named, reusable function so you could just call one of those here).

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/160dba9505eb98949a93c035ffbf676f3ac6bbd4/lms/static/js/header/header.js#L104-L105

These should be called on line 87 so they fire ASAP.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L149-L151

You can chain these with jQuery.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L136-L137

These should be called on line 110 so they fire ASAP.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/js/header/header.js#L170

This should be called on line 158 so it can fire ASAP.

AlasdairSwan commented:
In the Sass there is a lot of styling elements (e.g. https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/sass/_header.scss#L91) styles should be applied to classes instead of elements.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/static/sass/shared-v2/_navigation.scss#L24

Are there no z-index variables we could use?

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/header.html#L45

Should have ```aria-hidden="true"```

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/header.html#L56

Having the spans without classes makes this code very confusing.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/header.html#L69

This needs to be aria hidden

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/header.html#L119-L126

Was this previously in the header? @marcotuts do we still need a warning for users using Less than IE9?

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/user_dropdown.html#L26-L27

I don't think this will translate correctly.

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/lms/templates/header/user_dropdown.html#L33

Why set the tabindex?

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/themes/edx.org/lms/templates/legacy_header.html#L40

Needs to be aria hidden

AlasdairSwan commented:
https://github.com/edx/edx-platform/blob/3d79e3e05d6e7a8ff1f88c1f9f928c633247ce42/themes/edx.org/lms/templates/legacy_header.html#L247-L254

Same question re. the need for this.

HarryRein commented:
Can't respond directly to the comments, but addressed the feedback. Couple things that are a bit different in the fixing:

1) Gonna talk to product about the UI design before matching, feel that the edx.org menu is a bit thick but will leave the decision up to them.
2) Can't scope the mobile menu to the header because it lives outside
3) I'm not sold yet on the not styling elements idea - where does that come from? It feels worse to to add classes everywhere when scoped elements work just fine and the SASS guarantees specificity.
4) Gonna leave the z-index thing, that file will be removed anyways once the pattern library lands
5) Agreed that the span sequence is somewhat confusing but its clear enough once you check in the inspector what's going on - if you have a particular thing you think would make sense then we can add it but if not I'll just leave as is
6) About the translation for the "Dashboard for" - why don't you think that will work? It was coming from the old LMS header.
7) tabindex is set because it's a span and spans don't typically get into the tab flow

AlasdairSwan commented:
@andy-armstrong the JS scoping issue isn't a previous one, @cptvitamin mentioned that it has introduced a bug on production.

AlasdairSwan commented:
This is the bug https://openedx.atlassian.net/browse/EDUCATOR-1640

HarryRein commented:
The bug above was due to missing parentheses around the check for up or down arrow. Fixed

cptvitamin commented:
All ```<nav>``` elements, need an aria-label attr on them to distinguish them from other nav elements and landmark elements. Not in this PR, but here is one on the non-authenticated nav. It needs to be done on the authenticated one as well.
https://github.com/edx/edx-platform/blob/160dba9505eb98949a93c035ffbf676f3ac6bbd4/themes/edx.org/lms/templates/header/navbar-not-authenticated.html#L19

marcotuts commented:
Ok to ignore ie9 messaging, it has been a long time since we even supported ie11. :)

cptvitamin commented:
Is everyone OK with the removal of all the microdata/stuctured data markup in this implementation? I would think, especially if we plan on making more of these pages available in front of the paywall, that we would want that in there.

HarryRein commented:
jenkins run python

AlasdairSwan commented on lms/templates/header/navbar-not-authenticated.html:
```diff
@@ -18,7 +18,7 @@
   allow_public_account_creation = static.get_value('ALLOW_PUBLIC_ACCOUNT_CREATION', settings.FEATURES.get('ALLOW_PUBLIC_ACCOUNT_CREATION'))
 %>
-<nav class="nav-links">
+<nav class="nav-links" aria-label="Supplemental Links">
```
String needs to be wrapped for translations.

AlasdairSwan commented on lms/templates/header/user_dropdown.html:
```diff
@@ -28,11 +28,10 @@
-    <div class="user-dropdown" aria-expanded="false" aria-haspopup="true" aria-controls="user-menu">
-        <span class="sr">user menu</span>
-        <span class="fa fa-caret-down toggle-user-dropdown" aria-hidden="true" tabindex="0"></span>
+    <div class="toggle-user-dropdown" role="button" aria-label="Options Menu" aria-expanded="false" tabindex="0" aria-controls="user-menu">
```
String needs to be wrapped for translations.

AlasdairSwan replied:
@HarryRein is there a reason that this is a ```div``` with role ```button``` instead of just making it a ```button``` element?

HarryRein replied:
Another instance of avoiding the overrides that come with the button class - if you switch in the chrome inspector it has a bunch of things that we would need to add to our styling which made it a bit confusing.

AlasdairSwan commented on lms/templates/header/user_dropdown.html:
```diff
@@ -28,11 +28,10 @@
+        <span class="fa fa-caret-down" aria-hidden="true"></span>
     </div>
-    <div class="dropdown-user-menu" aria-label="More Options" role="menu" id="user-menu" tabindex="-1">
+    <div class="dropdown-user-menu hidden" aria-label="More Options" role="menu" id="user-menu" tabindex="-1">
```
String needs to be wrapped for translations.

AlasdairSwan commented on lms/templates/header/header.html:
```diff
@@ -42,79 +38,29 @@
+    <%include file="navbar-not-authenticated.html" args="online_help_token=online_help_token"/>
+  % endif
+</header>
+<div class="mobile-menu hidden" aria-label="More Options" role="menu" id="mobile-menu"></div>
```
String needs to be wrapped for translations.

AlasdairSwan commented on lms/templates/header/header.html:
```diff
@@ -42,79 +38,29 @@
 
+<header class="global-header ${'slim' if course else ''}">
+  <%include file="navbar-logo-header.html" args="online_help_token=online_help_token"/>
+  <div class="hamburger-menu" role="button" aria-label="Options Menu" aria-expanded="false" aria-controls="mobile-menu" tabindex="0">
```
String needs to be wrapped for translations.
Is there a reason that this is a ```div``` with ```role="button"``` instead of just a ```button``` element?

HarryRein replied:
It seemed like a simpler implementation than overriding all the button styling that gets tied to the element.

AlasdairSwan replied:
I'm fine with shipping this as is for now to get this out, but I would prefer to utilize the correct HTML elements and update Sass than force a div to act like a button. Because this is a pattern we use throughout the site (buttons that look like link) there would be value in adding a class to easily style buttons as links (not necessarily in this PR).

andy-armstrong replied:
There is a class `.btn-link` that works in at least some of the styles. It wouldn't take too much to get it to work in all three.

edx-webhook commented:
Thanks for the pull request, @HarryRein! I've created [OSPR-2202](https://openedx.atlassian.net/browse/OSPR-2202) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented:
@HarryRein Welcome to life after edX!  Pull requests that are still open will occasionally be found by the OSPR bot.  Since you don't have a valid current contributor agreement, it decides that you need an OSPR ticket created.

@AlasdairSwan Do we think this might someday merge, or should we close it?

AlasdairSwan commented:
@nedbat I will take a more in-depth look later today, but it's possible that this work was implemented in a different PR (https://github.com/edx/edx-platform/commit/63eeca73ce5102f4de16c33198379978f3cb1bf0)

AlasdairSwan commented:
@nedbat this PR can be closed because the work was already committed in a different PR (https://github.com/edx/edx-platform/commit/63eeca73ce5102f4de16c33198379978f3cb1bf0)
